src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/DeleteSearchParameterBehavior.cs

```diff
@@ -0,0 +1,54 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Messages.Delete;
+using Microsoft.Health.Fhir.Core.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Parameters
+{
+    public class DeleteSearchParameterBehavior<TDeleteResourceRequest, TDeleteResourceResponse> : IPipelineBehavior<TDeleteResourceRequest, TDeleteResourceResponse>
+        where TDeleteResourceRequest : DeleteResourceRequest
+    {
+        private ISearchParameterUtilities _searchParameterUtilities;
+        private IFhirDataStore _fhirDataStore;
+
+        public DeleteSearchParameterBehavior(ISearchParameterUtilities searchParameterUtilities, IFhirDataStore fhirDataStore)
+        {
+            EnsureArg.IsNotNull(searchParameterUtilities, nameof(searchParameterUtilities));
+            EnsureArg.IsNotNull(fhirDataStore, nameof(fhirDataStore));
+
+            _searchParameterUtilities = searchParameterUtilities;
+            _fhirDataStore = fhirDataStore;
+        }
+
+        public async Task<TDeleteResourceResponse> Handle(TDeleteResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TDeleteResourceResponse> next)
+        {
+            var deleteRequest = request as DeleteResourceRequest;
+            ResourceWrapper searchParamResource = null;
+
+            if (deleteRequest.ResourceKey.ResourceType.Equals(KnownResourceTypes.SearchParameter, StringComparison.Ordinal))
+            {
+                searchParamResource = await _fhirDataStore.GetAsync(deleteRequest.ResourceKey, cancellationToken);
+            }
+
+            var response = await next();
```

Comment:
>next [](start = 33, length = 4)

nit:
Part of me want to argue about order of things.
So if correctly understand there is two parts, storing search parameter in data store, and storing search parameter in memory of the server.
By having search parameter in memory, we apply it to all incoming data, and i think we should purge it from memory first and then try to remove it from server. If someone try to delete search parameter i would assume they don't want it to be applied anymore, so it sound a bit unnatural for me to remove it from memory as second step.
Also if I look into adding search parameter order is : add to data store, then add to memory.
Considering what delete operation is opposite of adding,  it doesn't feel like it should follow same order, and should be opposite.

Reply:
I had the same thought about order and considered the alternate approach.  Currently the call to DeleteSearchParameterStatusAsync uses the metadata stored in the SearchParameterDefinitionManager to then go and remove the status from the data store.  So, because of dependencies, it does need to happen in this order.

Reply:
Would it worth to add comment regarding order? I think we both agree it's a bit counterintuitive, so maybe it's worth to leave some information here.

---
In reply to: [560619433](https://github.com/microsoft/fhir-server/pull/1581#discussion_r560619433) [](ancestors = 560619433)